File: haskell/PrimeFactors.hs
module PrimeFactors (primeFactors) where

primeFactors :: Integer -> [Integer]
primeFactors x = primeFactorsInternal x 2
    where
        primeFactorsInternal x i
            | x == 1 = []
            | x < i * i = [x]
            | x `mod` i == 0 = i : primeFactorsInternal (x `div` i) i
            | otherwise = primeFactorsInternal x (i + 1)

--  vim: set et sts=4 sw=4 ts=8 :

File: scala/src/main/scala/PrimeFactors.scala
package com.crebma.primeFactors

object PrimeFactors {

  def factors(num: Int) : List[Int] = {
    factors(num, 2)
  }

  def factors(num: Int, candidate: Int) : List[Int] = {
    if (num == 1) {
      List[Int]()
    } else if (num % candidate == 0) {
      candidate :: factors(num / candidate, candidate)
    } else {
      factors(num, candidate + 1)
    }
  }
}

//  vim: set et sts=2 sw=2 ts=2 :
